File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.*;
class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        int answer = 0;
        int totalWei = 0;
        int truckNum = 0;
        Queue<Integer> bridge = new ArrayDeque<>(); 
        
        for(int i=0; i<bridge_length;i++){
            bridge.offer(0);
        }
        
        while(truckNum!=truck_weights.length){
            answer++;
        
            int in = truck_weights[truckNum];
            int out = bridge.poll();
        
            totalWei = totalWei + in - out;
            
            if(totalWei>weight){
                totalWei = totalWei - in; 
                bridge.offer(0);
                
            }else{
                bridge.offer(in);
                truckNum++;
            }
            
            
        }
        
        answer = answer + bridge_length;
        
        
        return answer;
    }
}

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
from collections import deque
def solution(bridge_length, weight, truck_weights):
    answer = 0
    time = 0
    bridge = deque()
    bridge.extend([0]*bridge_length)
    total_weight = 0
    while truck_weights:
        time += 1
        out = bridge.popleft()
        total_weight -= out
        if(total_weight+truck_weights[0]<=weight):
            truck = truck_weights.pop(0)
            bridge.append(truck)
            total_weight += truck   
        else: bridge.append(0)
    
    time +=bridge_length
    answer = time
        
        
    return answer